// @pyroscope/nodejs/dist/cjs/middleware/fastify.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const debug_1 = __importDefault(require("debug"));
const pyroscope_profiler_js_1 = require("../utils/pyroscope-profiler.js");
const pprof_1 = require("@datadog/pprof");
const log = (0, debug_1.default)('pyroscope');
async function collectProfile(profiler) {
    const profile = profiler.profile().profile;
    profiler.stop();
    return (0, pprof_1.encode)(profile);
}
async function collectProfileAfterMs(profiler, args, delayMs) {
    profiler.start(args);
    if (delayMs === 0) {
        return collectProfile(profiler);
    }
    return new Promise((resolve) => {
        setTimeout(() => {
            resolve(collectProfile(profiler));
        }, delayMs);
    });
}
function collectHeap() {
    const profiler = (0, pyroscope_profiler_js_1.getProfiler)();
    const heapProfilerArgs = profiler.heapProfiler.startArgs;
    const heapProfiler = profiler.heapProfiler.profiler;
    return collectProfileAfterMs(heapProfiler, heapProfilerArgs, 0);
}
function collectWall(ms) {
    const profiler = (0, pyroscope_profiler_js_1.getProfiler)();
    const wallProfilerArgs = profiler.wallProfiler.startArgs;
    const wallProfiler = profiler.wallProfiler.profiler;
    return collectProfileAfterMs(wallProfiler, wallProfilerArgs, ms);
}
async function heapHandler(request, reply) {
    log('Fetching Heap Profile');
    try {
        const profileBuffer = await collectHeap();
        reply.status(200).type('application/octet-stream').send(profileBuffer);
    }
    catch (error) {
        log('Error collecting Heap', error);
        reply.status(500).send({ error: 'Internal Server Error' });
    }
}
async function wallHandler(request, reply) {
    log('Fetching Wall Profile');
    try {
        const seconds = Number(request.query.seconds || 1);
        const profileBuffer = await collectWall(1000 * seconds);
        reply.status(200).type('application/octet-stream').send(profileBuffer);
    }
    catch (error) {
        log('Error collecting Wall', error);
        reply.status(500).send({ error: 'Internal Server Error' });
    }
}
const fastifyMiddleware = () => {
    const plugin = (fastify, _options, done) => {
        // Register route for heap profiling
        fastify.get('/debug/pprof/heap', heapHandler);
        // Register route for wall/CPU profiling
        fastify.get('/debug/pprof/profile', wallHandler);
        done();
    };
    return plugin;
};
exports.default = fastifyMiddleware;
