// @pyroscope/nodejs/dist/esm/middleware/fastify.js
import debug from 'debug';
import { getProfiler } from '../utils/pyroscope-profiler.js';
import { encode } from '@datadog/pprof';
const log = debug('pyroscope');
async function collectProfile(profiler) {
    const profile = profiler.profile().profile;
    profiler.stop();
    return encode(profile);
}
async function collectProfileAfterMs(profiler, args, delayMs) {
    profiler.start(args);
    if (delayMs === 0) {
        return collectProfile(profiler);
    }
    return new Promise((resolve) => {
        setTimeout(() => {
            resolve(collectProfile(profiler));
        }, delayMs);
    });
}
function collectHeap() {
    const profiler = getProfiler();
    const heapProfilerArgs = profiler.heapProfiler.startArgs;
    const heapProfiler = profiler.heapProfiler.profiler;
    return collectProfileAfterMs(heapProfiler, heapProfilerArgs, 0);
}
function collectWall(ms) {
    const profiler = getProfiler();
    const wallProfilerArgs = profiler.wallProfiler.startArgs;
    const wallProfiler = profiler.wallProfiler.profiler;
    return collectProfileAfterMs(wallProfiler, wallProfilerArgs, ms);
}
async function heapHandler(request, reply) {
    log('Fetching Heap Profile');
    try {
        const profileBuffer = await collectHeap();
        reply.status(200).type('application/octet-stream').send(profileBuffer);
    }
    catch (error) {
        log('Error collecting Heap', error);
        reply.status(500).send({ error: 'Internal Server Error' });
    }
}
async function wallHandler(request, reply) {
    log('Fetching Wall Profile');
    try {
        const seconds = Number(request.query.seconds || 1);
        const profileBuffer = await collectWall(1000 * seconds);
        reply.status(200).type('application/octet-stream').send(profileBuffer);
    }
    catch (error) {
        log('Error collecting Wall', error);
        reply.status(500).send({ error: 'Internal Server Error' });
    }
}
const fastifyMiddleware = () => {
    const plugin = (fastify, _options, done) => {
        // Register route for heap profiling
        fastify.get('/debug/pprof/heap', heapHandler);
        // Register route for wall/CPU profiling
        fastify.get('/debug/pprof/profile', wallHandler);
        done();
    };
    return plugin;
};
export default fastifyMiddleware;
